### ISSUE TITLE
contribution of NaN value doesn't work if the feature is numeric

### ISSUE THREAD

**Issue by sam94700** [MAIF/shapash#580]

**Comment by immu4989** [comment]
Happy to take a look at this . The issue title alone left a few possible
interpretations, so I tried to reproduce on the latest release and trace
the behaviour back to a specific code path. Posting findings before
opening a PR in case the maintainers want to weigh in on the proposed fix
direction.

### Reproducing on `v2.8.1`

```python
import numpy as np, pandas as pd
from sklearn.ensemble import HistGradientBoostingClassifier
from shapash import SmartExplainer

rng = np.random.default_rng(0)
n, n_nan = 60, 5
nf = np.concatenate([rng.normal(size=n - n_nan), [np.nan] * n_nan])
df = pd.DataFrame({"num_feat": nf, "noise": rng.normal(size=n)})
y = pd.Series(np.where(np.isnan(nf), 1, (nf > 0).astype(int)), name="target")

model = HistGradientBoostingClassifier(max_iter=100, random_state=0).fit(df, y)
xpl = SmartExplainer(model=model)
xpl.compile(x=df, y_target=y)

# SHAP contributions for the NaN-feature rows are computed correctly:
nan_idx = xpl.x_init.index[xpl.x_init["num_feat"].isna()]
print(xpl.contributions[1].loc[nan_idx, "num_feat"])
# 55..59 → ~2.31 each (the model learned a "missing-value" branch)

fig = xpl.plot.contribution_plot("num_feat")
scatter = [t for t in fig.data if t.type == "scatter" and t.mode == "markers"][0]
sx = np.asarray(scatter.x, dtype=float)
print(f"trace points: {len(sx)},  with x=NaN: {int(np.isnan(sx).sum())}")
# trace points: 60,  with x=NaN: 5
```

The 5 NaN-feature rows reach the figure with valid `y` (the SHAP value)
but with `x = NaN`. Plotly silently omits markers whose x is NaN, so
those rows are invisible in the rendered chart even though their
contribution is computed and would be informative — exactly the case the
title points to.

### Where it originates

Both branches of `SmartPlotter.contribution_plot` ([smart_plotter.py:586,
:609](https://github.com/MAIF/shapash/blob/v2.8.1/shapash/explainer/smart_plotter.py#L586))
hide NaN rows, but for different reasons:

- **`plot_scatter`** —
  [plots/plot_contribution.py:125-171](https://github.com/MAIF/shapash/blob/v2.8.1/shapash/plots/plot_contribution.py#L125-L171).
  `feature_values_array` is passed straight into `fig.add_scatter(x=...)`
  with NaNs intact. NaN is stripped at line 137 only for the KDE input,
  not for the rendered markers, so Plotly drops them.
- **`plot_violin`** —
  [plots/plot_contribution.py:290](https://github.com/MAIF/shapash/blob/v2.8.1/shapash/plots/plot_contribution.py#L290).
  `feature_values.value_counts()` defaults to `dropna=True`, so NaN
  never enters `xs`. The downstream `feature_values.iloc[:, 0] == c`
  filters (lines 323 / 343 / 363) also never match NaN rows, so they
  are excluded from every violin/scatter group.

For comparison, the categorical path *does* preserve missing values:
[`utils/transform.py:396`](https://github.com/MAIF/shapash/blob/v2.8.1/shapash/utils/transform.py#L396)
fills with `"missing"` and
[`utils/category_encoder_backend.py:105`](https://github.com/MAIF/shapash/blob/v2.8.1/shapash/utils/category_encoder_backend.py#L105)
uses the `"NaN"` label, so a categorical feature's NaN modality is
plotted as its own bucket. The asymmetry between the two paths matches
the title.

### Proposed fix direction

The contributions themselves are correct ,only the plot layer hides
them , so the fix can be local to `plots/plot_contribution.py`:

- For `plot_scatter`, render NaN feature rows as a separate marker trace
  pinned to a labeled "NaN" / "missing" tick placed just outside the
  numeric x-range, mirroring what categorical NaN already does.
- For `plot_violin`, switch to `value_counts(dropna=False)` and add a
  dedicated `feature_values.iloc[:, 0].isna()` group so NaN appears as
  its own modality on the x-axis.

Both keep the existing colour / hover behaviour and avoid touching the
explanation pipeline.

### Context

I ran into this on production tabular data (HR risk modelling on
XGBoost / RandomForest with SHAP based explanations) where the
"missing" branch of a tree often carries real signal ,being able to
see those rows on the contribution plot matters.

Happy to put up a PR along the lines above if that direction looks
reasonable, or to adjust if you'd prefer a different approach (e.g.
keeping NaN folded into the density curve rather than as a separate
tick). Tested against `v2.8.1` on Python 3.9, pandas 2.3.3, numpy 2.0.2.

**Comment by guillaume-vignal** [comment]
Thanks a lot for the detailed investigation and write-up — this is extremely helpful.

You're absolutely right: the issue was not clearly described initially, and your analysis helped narrow down the actual problem very precisely.

Feel free to open a PR. Thanks again for taking the time to investigate this.
